src/praisonai-train/praisonai_train/data/benchmark.py

```
@@ -0,0 +1,237 @@
+    still contributes its latency but counts as ``ok=False`` with 0 tokens, so
+    error rates show up as a lower success count rather than crashing the run.
+    Tokens are only reported for successful (``ok=True``) responses so token
+    throughput and average-output-tokens stay consistent with the success count.
+    ``json_mode`` toggles the ``response_format`` hint; disable it for endpoints
+    that don't support JSON mode. This is the single network seam — tests
+    monkeypatch it.
+    """
+    req = build_chat_request(cfg, messages, json_mode=json_mode)
+    t0 = time.perf_counter()
+    try:
+        with urllib.request.urlopen(req, timeout=timeout) as resp:
```

Ruff (0.15.21)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)